pkg/lassie/lassie.go

```diff
@@ -0,0 +1,60 @@
+package lassie
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/filecoin-project/lassie/internal"
+	"github.com/filecoin-project/lassie/pkg/retriever"
+	"github.com/filecoin-project/lassie/pkg/types"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-graphsync/storeutil"
+	carblockstore "github.com/ipld/go-car/v2/blockstore"
+)
+
+type Lassie struct {
+	timeout time.Duration
+}
+
+func NewLassie(timeout time.Duration) *Lassie {
+	return &Lassie{
+		timeout: timeout,
+	}
+}
+
+func (l *Lassie) Fetch(ctx context.Context, rootCid cid.Cid, outfile *os.File) (*types.RetrievalStats, types.RetrievalID, error) {
+	var parentOpener = func() (*carblockstore.ReadWrite, error) {
+		return carblockstore.OpenReadWriteFile(outfile, []cid.Cid{rootCid})
+	}
+
+	var blockCount int
+	var byteLength uint64
+	putCb := func(putCount int, putBytes int) {
+		blockCount += putCount
+		byteLength += uint64(putBytes)
+	}
+	bstore := internal.NewPutCbBlockstore(parentOpener, putCb)
+
+	linkSystem := storeutil.LinkSystemForBlockstore(bstore)
+
+	var ret *retriever.Retriever
+	ret, err := internal.SetupRetriever(ctx, l.timeout)
+	if err != nil {
+		return nil, types.RetrievalID{}, err
+	}
```

Comment:
this should persist across `Lassie` instances, you either need a `Setup()` / `Start()` function to do this or build it in to `NewLassie()`. Multiple calls through `Fetch` to the same instance should use the same `Retriever` instance.

Reply:
I would go one further -- I'm not sure the value of this over SetupRetriever when you add everything up. I imagine the HTTP server to just have a copy of the retriever and then to include the few extra lines of code there.

Reply:
A context is required when instantiating a `Retriever` for the `EventManager`. We would use the daemon cli context for the `Retriever` instantiation. And then for the actual `Retrieve()` call, we'd use the request context. After talking with you guys offline, it sounds like this is right. The request context will also be cancelled if the server context is cancelled, so everything is fine with either scenario.